[PL-128135] Add s3 gateway on rgw servers

Ma27 opened the pull request:
@flyingcircusio/release-managers

## Release process

- [ ] Created changelog entry using `./changelog.sh`

## PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [ ] ticket state set to *Pull request ready*
- [ ] if ticket is more urgent than within the next few days, directly contact a member of the Platform team
<!---->
- [ ] set urgency and risk labels
- [ ] ensure the merge bot has determined a merge date
- [ ] ensure all checks are green
- [ ] get a review from a colleague

## Design notes

- [ ] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [ ] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [ ] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
- [ ] Security requirements tested? (EVIDENCE)

Commit:
roles/ceph/rgw: add s3/object gateway on rgw machines

PL-128135

We no longer let radosgw listen on port 80, but on port 7480 instead and
let the reverse proxy in front listen on port 80.

The haproxy in between uses other rgws belonging to the same cluster as
backup servers.

TLS termination is done with nginx. The ACME challenge is DNS-01 since
we usually have multiple RGW instances.

Commit:
fc-ceph: code to summarize daily traffic stats of the object buckets & GC old data

PL-128135

Commit:
roles/ceph/mon: add timer to run accounting regularly

PL-128135

Commit:
roles/ceph/rgw: sensu check to ensure accounting is actually happening

PL-128135

Commit:
tests/ceph-nautilus: check that integration for traffic accounting works

PL-128135

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -21,18 +21,33 @@ let
 
   cephPkgs = fclib.ceph.mkPkgs role.cephRelease;
 
+  radosListenPort = toString 7480;
```
I *think* we need to leave 80 in here for pre-signed URLs?

Ma27 replied:
Do you happen to have a reproducer for the old issue? Because I cannot reproduce on dev (and the tests are fine either):

```
$ cat .aws/credentials
[dev2]
endpoint_url = https://objects.dev.fcio.net
<redacted>
$ aws --profile dev2 s3 ls s3://test/
[...]
2025-07-21 13:21:18         12 testfile
$ aws --profile dev2 s3 presign s3://test/testfile
https://objects.dev.fcio.net/test/testfile?AWSAccessKeyId=03KAF2L1WU7PHIH0H66M&Signature=Oh1zNWA0TnHbdNCPerhlEEG%2FqHU%3D&Expires=1755084588
$ curl "https://objects.dev.fcio.net/test/testfile?AWSAccessKeyId=03KAF2L1WU7PHIH0H66M&Signature=Oh1zNWA0TnHbdNCPerhlEEG%2FqHU%3D&Expires=1755084588"
Hello World
```

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -21,18 +21,33 @@ let
 
+  radosListenPort = toString 7480;
+  haproxyPort = toString 7475;
+  vhostName = lib.concatStringsSep "." (
```
I'd rather have this being a proper option where the computed value is the default and we can completely override this. Otherwise this appears to be a potential "let trap".

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -21,18 +21,33 @@ let
     rgwMimeTypesFile = "${pkgs.mime-types}/etc/mime.types";
     debugRados = "1 5";
-    rgwFrontends = "beast port=80";
+    rgwFrontends = "beast port=${radosListenPort}";
```
please bind explicity to srv (but also keep the firewall!)

>  Sets the listening address in the form address[:port], where the address is an IPv4 address string in dotted decimal form, or an IPv6 address in hexadecimal notation surrounded by square brackets. Specifying a IPv6 endpoint would listen to v6 only. The optional port defaults to 80 for endpoint and 443 for ssl_endpoint. Can be specified multiple times as in endpoint=[::1] endpoint=192.168.0.100:8000.

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -144,37 +235,22 @@ in
-          sto = fclib.network.sto;
         in
-        lib.mkMerge [
-          (lib.mkOrder 700 ''
```
those might be needed if we have to enable the port redirections again

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -210,7 +286,11 @@ in
       flyingcircus.services.sensu-client.checks = {
         radosgw_probe_object = {
```
lets name this rados_gw_probe_object_nginx

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -210,7 +286,11 @@ in
-          command = "check_http -u /rgw-monitoring/probe -H localhost -m 1000000:1500000 -w 5 -c 10";
+          command = "check_http -u /rgw-monitoring/probe -H localhost:7480 -m 1000000:1500000 -w 5 -c 10";
+        };
+        radosgw_probe_object_gw = {
```
lets name this radosgw_probe_object_direct

ctheune commented on nixos/roles/ceph/rgw.nix:
```diff
@@ -271,6 +351,32 @@ in
+            with open("/var/lib/fc-ceph-s3-accounting/s3-accounting-state", "r") as f:
+                data = json.load(f)
+
+            THRESH_WARN = timedelta(days=1)
```
6 hours warning
12 hours critical.

ctheune commented on pkgs/fc/ceph/src/fc/ceph/logs.py:
```diff
@@ -50,3 +55,40 @@ def slowreq_histogram(self, include, exclude, first_n, filenames):
                 break
             n += 1
+
+    def account_s3_traffic(self, state_file: Path, enc_file: Path):
```
lets gate the accounting timer behind a feature flag so in case this explodes and we have to adjust things

ctheune replied:
This is also a potential RAM bottleneck for processing if we have to parse the json completely. At least this is split by bucket so that helps, but we might have to switch to an event-based json parser (bleh) in the future ... leave as is for now.

ctheune commented on pkgs/fc/ceph/src/fc/ceph/util/opslog.py:
```diff
@@ -0,0 +1,130 @@
+
+    def gc_log_objects(self):
+        day = self.opslog_state.last_gced_day + timedelta(days=1)
+        end = self.opslog_state.last_processed_datetime.date() - timedelta(
```
lets have a small guard that ensures that end is always at most yesterday and never today or in the future.

ctheune reviewed (changes requested):
Definitely on the right track, see the comments.

Commit:
tests/ceph-nautilus: assert min_pgs instead of pgs

PL-128135

Upon working on PL-128135, I encountered a case where in the early setup
either 64 or 320 pgs were present breaking the test. Since this is
timing-related, only check that _at least_ 64 pgs are free.

Commit:
fc-ceph: filter out internal traffic on s3 accounting

PL-128135

leona-ya commented:
continued in #2142
